reward_extraction/behavior_cloning/behavior_cloning.py

```diff
@@ -0,0 +1,191 @@
+
+        return x
+    
+
+def data(SCALE) -> Tuple[DataLoader, DataLoader]:
+    expert_data = RoboDemoDset('/home/xarm/Documents/JacobAndDavin/test/rewardlearning-robot/data/demos/test/demos.hdf', read_only_if_exists=True)
+    state = []
+    targets = []
+    for traj in expert_data:
+        joint_poses = traj['eef_pose']
+        # delta_targets = [] # np.zeros((len(joint_poses)-1, 3))
+        for i in range(0, len(joint_poses) -1):
+            if np.sum(joint_poses[i+1] - joint_poses[i]) == 0:
+                continue
+            state.append(joint_poses[i] / SCALE)
+            targets.append(joint_poses[i+1] / SCALE - joint_poses[i] / SCALE)
+            # targets.append(joint_poses[i]/100)
+        # print(joint_poses[:10])
+        # print(delta_targets[:10])
+
+        # targets.append(targets)
+        
+    state = np.array(state)
+    targets = np.array(targets)
+    perm = torch.randperm(len(state))
+    state = torch.Tensor(state)
+    targets = torch.Tensor(targets)
+    state = state[perm]
+    targets = targets[perm]
+    test_idx = int(len(state) * 0.8)
+    train_x = state[:test_idx]
+    train_y = targets[:test_idx]
+    test_x = state[test_idx:]
+    test_y = targets[test_idx:]
+
+    train_loader = DataLoader(BCDataset(train_x, train_y), batch_size=128, shuffle=True)
+    test_loader = DataLoader(BCDataset(test_x, test_y), batch_size=128, shuffle=True)
+    return train_loader, test_loader
+
+
+
+class BCDataset(Dataset):
+    def __init__(self, x, y) -> None:
+        # super().__init__()
+        self.x = x
+        self.y = y
+    def __len__(self):
+        return(len(self.x))
+
+    def __getitem__(self, idx):
+        return self.x[idx], self.y[idx]
+
+def one_train_iter(net, optimizer, train_loader):
+    critirion = nn.MSELoss()
+    net.train()
+    losses = []
+    for data, labels in train_loader:
+        data, labels = data.to(device), labels.to(device)
+        optimizer.zero_grad()
+        preds = net(data)
+        loss = critirion(preds, labels)
+        losses.append(loss.item())
+        loss.backward()
+        optimizer.step()
+    print(f'Average train loss is {np.mean(losses)}')
+    return (np.mean(losses))
+
+def eval(net, test_loader):
+    critirion = nn.MSELoss()
+    with torch.no_grad():
+        losses = []
+        for data, labels in test_loader:
+            data, labels = data.to(device), labels.to(device)
+            preds = net(data)
+            # print(preds)
+            # print(labels)
+            loss = critirion(preds, labels)
+            loss = loss.item()
+            losses.append(loss)
+        print(f'Average test loss is {np.mean(losses)}')
+
+
+def try_until_estop(model, SCALE):
+    env = XArmTaskSpaceEnv(
+        control_frequency_hz=10,
+        use_camera=True,
+        # TODO: create some sort of button pressing mechanism to open and close the gripper,
+        random_reset_home_pose=False
+    )
+    obs = env.reset()
+    model.eval()
+    model.double()
```

Comment:
generally leaving models with float precision is fine instead of doubles

Reply:
we ran into error when we left it at float: `expected scaler type double` but found float. Maybe we made an error somewhere I will look into it thanks